specs/_features/eip7251/beacon-chain.md

```diff
@@ -0,0 +1,827 @@
+    per_epoch_churn = get_activation_exit_churn_limit(state)
+    # New epoch for exits.
+    if state.earliest_exit_epoch < earliest_exit_epoch:
+        state.earliest_exit_epoch = earliest_exit_epoch
+        state.exit_balance_to_consume = per_epoch_churn
+
+    if exit_balance <= state.exit_balance_to_consume:
+        # Exit fits in the current earliest epoch.
+        state.exit_balance_to_consume -= exit_balance
+    else:
+        # Exit doesn't fit in the current earliest epoch.
+        balance_to_process = exit_balance - state.exit_balance_to_consume
+        additional_epochs, remainder = divmod(balance_to_process, per_epoch_churn)
+        state.earliest_exit_epoch += additional_epochs + 1
+        state.exit_balance_to_consume = per_epoch_churn - remainder
+
+    return state.earliest_exit_epoch
+```
+
+#### New `compute_consolidation_epoch_and_update_churn`
+
+```python
+def compute_consolidation_epoch_and_update_churn(state: BeaconState, consolidation_balance: Gwei) -> Epoch:
+    earliest_consolidation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
+    per_epoch_consolidation_churn = get_consolidation_churn_limit(state)
+    # New epoch for consolidations.
+    if state.earliest_consolidation_epoch < earliest_consolidation_epoch:
+        state.earliest_consolidation_epoch = earliest_consolidation_epoch
+        state.consolidation_balance_to_consume = per_epoch_consolidation_churn
+
+    if consolidation_balance <= state.consolidation_balance_to_consume:
+        # Consolidation fits in the current earliest consolidation epoch.
+        state.consolidation_balance_to_consume -= consolidation_balance
+    else:
+        # Consolidation doesn't fit in the current earliest epoch.
+        balance_to_process = consolidation_balance - state.consolidation_balance_to_consume
+        additional_epochs, remainder = divmod(balance_to_process, per_epoch_consolidation_churn)
+        state.earliest_consolidation_epoch += additional_epochs + 1
+        state.consolidation_balance_to_consume = per_epoch_consolidation_churn - remainder
+
+    return state.earliest_consolidation_epoch
+```
+
+#### Updated `slash_validator`
+
+```python
+def slash_validator(state: BeaconState,
+                    slashed_index: ValidatorIndex,
+                    whistleblower_index: ValidatorIndex=None) -> None:
+    """
+    Slash the validator with index ``slashed_index``.
+    """
+    epoch = get_current_epoch(state)
+    initiate_validator_exit(state, slashed_index)
+    validator = state.validators[slashed_index]
+    validator.slashed = True
+    validator.withdrawable_epoch = max(validator.withdrawable_epoch, Epoch(epoch + EPOCHS_PER_SLASHINGS_VECTOR))
+    state.slashings[epoch % EPOCHS_PER_SLASHINGS_VECTOR] += validator.effective_balance
+    slashing_penalty = validator.effective_balance // MIN_SLASHING_PENALTY_QUOTIENT_EIP7251  # [Modified in EIP7251]
+    decrease_balance(state, slashed_index, slashing_penalty)
```

Comment:
It looks like the whistleblower reward part is missing/removed. Is that intentional?

Reply:
A rough consensus is to set `MIN_SLASHING_PENALTY_QUOTIENT_EIP7251` and whistleblower reward to the same value to avoid minting ETH. I am planning to propose this change in the next few days